### ISSUE TITLE
-Z1 expected to find a string with 1-char edit distance from pattern

### ISSUE THREAD

**Issue by korziner** [Genivia/ugrep#206]
(echo Степан;echo Стефан) | ugrep -niZ1 Степан
1:Степан

Expected output:
1:Степан
2:Стефан

Workaround is to increase to Z2:
(echo Степан;echo Стефан) | ugrep -niZ2 Степан
1:Степан
2:Стефан

Adding ~ did not help: 
(echo Степан;echo Стефан) | ugrep -niZ~1 Степан
1:Степан

ugrep -V
ugrep 3.6.0 x86_64-pc-linux-gnu +sse2 +pcre2_jit +zlib +bzip2 +lzma +zstd 
Reproducible with 3.7.9

Reference tool works as expected:
(echo Степан;echo Стефан)|tre-agrep -n --max-errors=1 Степан
1:Степан
2:Стефан
which tre-agrep|xargs ldd
	linux-vdso.so.1 =>  (0x00007ffe73776000)
	libtre.so.5 => /usr/lib/x86_64-linux-gnu/libtre.so.5 (0x00007fe62488d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe6244c3000)
	/lib64/ld-linux-x86-64.so.2

**Comment by genivia-inc** [comment]
It should work, but something isn't quite right I'm afraid. I will take a look to fix this.

The one-char edit distance metric takes Unicode characters into account. It this works with an 'x':
~~~
( echo Степан ; echo Стефан ) | src/ugrep -Z1 Стеxан
~~~
and should work with any other Unicode character.

**Comment by genivia-inc** [comment]
I found the problem in the FuzzyMatcher's backtracking logic and will update ugrep after some more testing.

Thanks for reporting this issue.

**Comment by genivia-inc** [comment]
A quick update.

The problem is fixed. The problem occurred when the fuzzy matcher hit a middle UTF-8 multibyte code when backtracking, backtracking on the wrong middle position instead of the beginning multibyte to backtrack. That can happen in some cases. I've added more unit tests to the updated [FuzzyMatcher](https://github.com/Genivia/FuzzyMatcher) sub-project to check all possibilities to prevent this issue from happening.

I also added a new `BIN` flag to the FuzzyMatcher sub-project. This flag fuzzy-matches ASCII/binary only, not full Unicode. This is useful when ugrep is invoked with the `-U` (`--binary`) flag to search binary data, which then. sets the FuzzyMatcher `BIN` flag. I think that makes ugrep a better search engine for binary fuzzy searching when we don't want to match UTF-8 multi byte sequences, because these may not be valid in binary files.

The ugrep update will be released soon.
